### PR TITLE
Property signatures

### PR DESCRIPTION
**Pull request by kensens**
Implementation of property signatures.

also fixes some minor bugs that cause flaky tests

### CODE REVIEW

File: crossbeam/property_signatures/property_signatures.py [hunk 1]
```diff
@@ -0,0 +1,474 @@
...
+  arity = value.num_free_variables
+  assert arity > 0
+  num_examples = value.num_examples
+  io_pairs_per_example = [[] for _ in range(num_examples)]
+  inputs_to_try = sum(VALUES_TO_TRY.values(), [])
+  for inputs_list in itertools.product(inputs_to_try, repeat=arity):
+    for lambda_fn, io_pairs in zip(value.values, io_pairs_per_example):
+      try:
+        result = lambda_fn(*inputs_list)
+        if result is not None:
```
Comment (Hanjun-Dai) [inline crossbeam/property_signatures/property_signatures.py:337]: minor: maybe we can also record down the inapplicable inputs_list (or the types) and serve as auxiliary features.

File: crossbeam/property_signatures/property_signatures.py [hunk 2]
```diff
@@ -0,0 +1,474 @@
...
+    value_module.OutputValue([-1, -2]),
+    fixed_length=True))
+
+
+def _property_signature_concrete_value(
+    value: value_module.Value,
+    output_value: value_module.Value,
+    fixed_length: bool = True) -> List[Tuple[float, bool, bool, float]]:
+  """Returns a property signature for a value w.r.t. a set of I/O examples."""
+  assert value.num_free_variables == 0
```
Comment (Hanjun-Dai) [inline crossbeam/property_signatures/property_signatures.py:319]: minor: also need `assert value.num_examples==output_value.num_examples`

**Review by Hanjun-Dai — APPROVED** [review]
LGTM! some minor comments but we can come back to those later.

File: crossbeam/property_signatures/property_signatures.py (hunk 2 above)
Comment (kensens) [reply crossbeam/property_signatures/property_signatures.py:319]: actually this isn't always true. constants (and expressions that are effectively constant) have num_examples==1 and we "broadcast" the single example to fit the multiple examples when combining/comparing to other values with higher num_examples

File: crossbeam/property_signatures/property_signatures.py (hunk 1 above)
Comment (kensens) [reply crossbeam/property_signatures/property_signatures.py:337]: the types end up participating in the property signature (through _type_property) already. do you mean we should collect them explicitly as an additional return value?

Comment (Hanjun-Dai) [reply crossbeam/property_signatures/property_signatures.py:337]: I was thinking probably the way it fails (either throwing error or returning none) might indicate something, but anyway this is not in need at the moment.

File: crossbeam/property_signatures/property_signatures.py (hunk 2 above)
Comment (Hanjun-Dai) [reply crossbeam/property_signatures/property_signatures.py:319]: got it!

**Pull request merged**